Reduce number of implicit meta tasks

mkrizek opened the pull request:
##### SUMMARY
This greatly reduces run time on large inventories since meta tasks are
executed in the main process sequentially and just executing them is expensive.

This change avoids running the following implicit meta tasks:
  * ``flush_handlers`` on hosts where no handlers are notified
  * ``noop`` for the linear strategy's lockstep, instead hosts that are
    not executing the current task are just not part of the current host loop

A playbook consiting of two simple plays both running on ~6000 hosts
runs in:
devel: 37s
this PR: 1.3s

ci_complete
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE

<!--- Pick one below and delete the rest -->

- Bugfix Pull Request

bcoca commented on lib/ansible/plugins/strategy/linear.py:
```diff
@@ -90,8 +83,6 @@ def _get_next_task_lockstep(self, hosts, iterator):
                 iterator.set_state_for_host(host.name, state)
                 host_tasks.append((host, task))
-            else:
-                host_tasks.append((host, noop_task))
```
won't this break host task count parity/lockstep?

mkrizek replied:
I don't think so. I figured that instead of executing `noop` meta tasks the host can just be ignored for the current host loop iteration, which would actually be no op. The host would be again included in the host/task loop once its task is the "current" one. According to my testing and CI this does not break the lockstep.

mkrizek committed:
Reduce number of implicit meta tasks

This greatly reduces run time on large inventories since meta tasks are
executed in the main process sequentially and just executing them is expensive.

This change avoids running the following implicit meta tasks:
  * ``flush_handlers`` on hosts where no handlers are notified
  * ``noop`` for the linear strategy's lockstep, instead hosts that are
    not executing the current task are just not part of the current host loop

A playbook consiting of two simple plays both running on ~6000 hosts
runs in:
devel: 37s
this PR: 1.3s

ci_complete

sivel commented:
Should we pull in the change for `_evaluate_conditional` too, where we short circuit the default of `if not task.when:`?

Or is there still a need for https://github.com/ansible/ansible/pull/82252 ?

s-hertel committed:
short-circuit when eval for meta tasks

webknjaz commented on lib/ansible/executor/play_iterator.py:
```diff
@@ -447,6 +447,24 @@ def _get_next_task_from_state(self, state, host):
+                        # saved in `handler_notifications` and handlers are notified directly
+                        # to prevent duplicate handler runs, so check whether any handler
+                        # is notified
+                        and not any(bool(h.notified_hosts) for h in self.handlers)
```
Would this read better with less nesting?
```suggestion
                        and all(not h.notified_hosts for h in self.handlers)
```

mkrizek committed:
Update lib/ansible/executor/play_iterator.py

Co-authored-by: Sviatoslav Sydorenko (Святослав Сидоренко) <wk.cvs.github@sydorenko.org.ua>

mkrizek commented:
> A playbook consisting of two simple plays both running on ~6000 hosts runs in:
> devel: 37s
> this PR: 1.3s

The above plays have no handlers. I wanted to see how negatively the "no notifications" check impacted the performance and added 200 (two hundred) handlers to both plays. The running time increased from ~1.3s to 2.0s - 2.15s (measured with `time`).
